### PR TITLE
Exceptions Review

### PR DESCRIPTION
**Pull request by heidijinxujia**
Question,
In which situation, the catch of checked exceptions need to call returnToCaller()? (Compare Line 213 with Line 446)
Thanks, :)！

### CODE REVIEW

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 1]
```diff
@@ -127,7 +127,7 @@ public void onReceive(Context context, Intent intent) {
             Logger.v(TAG, "ActivityBroadcastReceiver onReceive");
 
             if (intent.getAction().equalsIgnoreCase(AuthenticationConstants.Browser.ACTION_CANCEL)) {
-                try {
+            //  try {
```
Comment (RandalliLama) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: I assume you are going to remove these lines.  Would you mind just doing that.  It clutters code reviews to see things that are in progress.

File: src/src/com/microsoft/aad/adal/AuthenticationContext.java [hunk 2]
```diff
@@ -1153,6 +1153,16 @@ private AuthenticationResult acquireTokenAfterValidation(CallbackHandler callbac
             AuthenticationResult result = null;
             request.setVersion(getVersionName());
             request.setBrokerAccountName(request.getLoginHint());
+            
+            //check the redirectUri before sending the request
+            //if the redirectUri from the client does not match the valid redirectUri
+            // the client app would not jump to the login page
```
Comment (RandalliLama) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: We only want to do this is skipBroker is false.

Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationContext.java]: This is under the case canSwithToBroker=true. But this code looks different compared to another PR for checking redirect uri @heidijinxujia .

**Comment by weijjia** [comment]
Can we get the travis issue resolved?

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 3]
```diff
@@ -657,9 +658,10 @@ public void alias(String alias) {
                         request.proceed(privateKey, certChain);
                         
                     } catch (KeyChainException e) {
-                        Log.e(TAG, "KeyChain exception", e);
+                        Log.e(TAG + ".onReceivedClientCertRequest()", "The alias of the desired certificate chain is valid, but have problems on accessing it.", e);
                     } catch (InterruptedException e) {
-                        Log.e(TAG, "InterruptedException exception", e);
+                    	//KeyChain.bindAsUser() throws InterruptedException
```
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: make sure we get all the commented code removed.

File: src/src/com/microsoft/aad/adal/AuthenticationActivity.java [hunk 4]
```diff
@@ -215,7 +211,10 @@ protected void onCreate(Bundle savedInstanceState) {
             mStartUrl = oauth.getCodeRequestUrl();
             mQueryParameters = oauth.getAuthorizationEndpointQueryParameters();
         } catch (UnsupportedEncodingException e) {
-            Log.d(TAG, e.getMessage());
+        	//checked exceptions
```
Comment (weijjia) [inline src/src/com/microsoft/aad/adal/AuthenticationActivity.java]: Make sure you get this removed. 
For this case, AuthenticationActivity is called in AuthenticationContext, we need to make sure we return the error back to AuthenticationContext.

**Comment by weijjia** [comment]
Make sure you get the conflicts resolved. Also, we should have more exception cases need to be covered, are you going to send them in multiple reviews?

**Pull request closed without merging**